crates/sdk/src/network.rs

```diff
@@ -0,0 +1,292 @@
+use crate::provers::SP1VerificationError;
+use crate::request::ProofRequest;
+
+use crate::network_v2::FulfillmentStrategy;
+use crate::network_v2::DEFAULT_PROVER_NETWORK_RPC;
+use crate::network_v2::{Error, RequestId, VerifyingKeyHash};
+use crate::verify;
+use crate::{network_v2::NetworkClient, proof::SP1ProofWithPublicValues};
+
+pub struct NetworkProver {
+    prover: SP1Prover<DefaultProverComponents>,
+    network_client: NetworkClient,
+}
+
+pub struct NetworkProverBuilder {
+    rpc_url: Option<String>,
+    private_key: Option<String>,
+}
+
+impl NetworkProver {
+    pub fn new(rpc_url: String, private_key: String) -> Self {
+        Self {
+            prover: SP1Prover::new(),
+            network_client: NetworkClient::new(&private_key).with_rpc_url(rpc_url),
+        }
+    }
+
+    pub fn with_mode(mut self, mode: Mode) -> Self {
+        self.network_client.mode = mode.into();
+        self
+    }
+
+    pub fn timeout_secs(mut self, secs: u64) -> Self {
+        self.network_client.timeout_secs = Some(secs);
+        self
+    }
+
+    pub fn cycle_limit(mut self, limit: u64) -> Self {
+        self.network_client.cycle_limit = Some(limit);
+        self
+    }
+
+    pub fn skip_simulation(mut self, skip: bool) -> Self {
+        self.network_client.skip_simulation = skip;
+        self
+    }
+
+    pub fn strategy(mut self, strategy: FulfillmentStrategy) -> Self {
+        self.network_client.fulfillment_strategy = Some(strategy);
+        self
+    }
+
+    #[allow(clippy::must_use_candidate)]
+    fn get_cycle_limit(
+        &self,
+        elf: &[u8],
+        stdin: &SP1Stdin,
+        cycle_limit: Option<u64>,
+        skip_simulation: bool,
+    ) -> Result<u64, Error> {
+        todo!()
+    }
+
+    /// Registers a program if it is not already registered.
+    pub async fn register_program(
+        &self,
+        vk: &SP1VerifyingKey,
+        elf: &[u8],
+    ) -> Result<VerifyingKeyHash> {
+        todo!()
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub async fn request_proof(
+        &self,
+        vk_hash: &VerifyingKeyHash,
+        stdin: &SP1Stdin,
+        version: &str,
+        mode: Mode,
+        strategy: FulfillmentStrategy,
+        timeout_secs: u64,
+        cycle_limit: u64,
+    ) -> Result<RequestId> {
+        todo!()
+    }
+
+    pub async fn wait_proof<P: DeserializeOwned>(
+        &self,
+        request_id: &RequestId,
+        timeout_secs: u64,
+    ) -> Result<P, Error> {
+        todo!()
+    }
+
+    pub fn prove_with_options<'a>(
+        &'a self,
+        pk: &'a SP1ProvingKey,
+        stdin: SP1Stdin,
+    ) -> NetworkProofRequest<'a> {
+        NetworkProofRequest::new(self, pk, stdin)
+    }
+
+    pub fn builder() -> NetworkProverBuilder {
+        NetworkProverBuilder::new()
+    }
+}
+
+impl NetworkProverBuilder {
+    pub fn new() -> Self {
+        Self { rpc_url: None, private_key: None }
+    }
+
+    pub fn with_rpc_url(mut self, url: String) -> Self {
```

Comment:
nit: `rpc_url`

Reply:
Is that better? one idea is that we might have a `rpc_url` getting func that returns the value, so maybe it is good to keep `with_` for the setter

Reply:
updated